diff --git a/llvm/runtimes/CMakeLists.txt b/llvm/runtimes/CMakeLists.txt
index 57db5c378..c7cd7d9d5 100644
--- a/llvm/runtimes/CMakeLists.txt
+++ b/llvm/runtimes/CMakeLists.txt
@@ -427,7 +427,7 @@ ${error} Set RUNTIMES_BUILD_ALLOW_DARWIN to allow a single darwin triple.")
                                         -DCMAKE_CXX_COMPILER_WORKS=ON
                                         -DCMAKE_ASM_COMPILER_WORKS=ON
                                         ${RUNTIMES_CMAKE_ARGS}
-                             PASSTHROUGH_PREFIXES LLVM_ENABLE_RUNTIMES
+                             PASSTHROUGH_PREFIXES LLVM_ENABLE_RUNTIMES CMAKE_MT
                                                   ${ARG_PREFIXES}
                              EXTRA_TARGETS ${extra_targets}
                                            ${test_targets}
