diff --git a/utils/build.ps1 b/utils/build.ps1
index 5de3ce1cef3..3a11840b9d1 100644
--- a/utils/build.ps1
+++ b/utils/build.ps1
@@ -66,6 +66,8 @@ The version number of the Windows SDK to be used.
 Overrides the value resolved by the Visual Studio command prompt.
 If no such Windows SDK is installed, it will be downloaded from nuget.
 
+.PARAMETER VCVarsVersion
+
 .PARAMETER IncludeDS2
 Include the ds2 remote debug server in the SDK.
 This component is currently only supported in Android builds.
@@ -146,6 +148,7 @@ param
   [string] $AndroidNDKVersion = "r27c",
   [ValidatePattern("^\d+\.\d+\.\d+(?:-\w+)?")]
   [string] $WinSDKVersion = "",
+  [string] $VCVarsVersion = "",
   [switch] $Android = $false,
   [switch] $SkipBuild = $false,
   [switch] $SkipPackaging = $false,
@@ -835,6 +838,10 @@ function Invoke-VsDevShell([Hashtable] $Platform) {
     $DevCmdArguments += " -winsdk=$WinSDKVersion"
   }
 
+  if ($VCVarsVersion) {
+    $DevCmdArguments += " -vcvars_ver=$VCVarsVersion"
+  }
+
   if ($ToBatch) {
     Write-Output "call `"$VSInstallRoot\Common7\Tools\VsDevCmd.bat`" $DevCmdArguments"
   } else {
@@ -1910,6 +1917,10 @@ function Build-mimalloc() {
     }
   }
 
+  if ($WinSDKVersion) {
+    Write-Host "Settingg WindowsTargetPlatformVersion to $WinSDKVersion"
+    $MSBuildArgs += "-p:WindowsTargetPlatformVersion=$WinSDKVersion"
+  }
   Invoke-Program $msbuild "$SourceCache\mimalloc\ide\vs2022\mimalloc-lib.vcxproj" @MSBuildArgs "-p:IntDir=$BinaryCache\$($Platform.Triple)\mimalloc\mimalloc\"
   Invoke-Program $msbuild "$SourceCache\mimalloc\ide\vs2022\mimalloc-override-dll.vcxproj" @MSBuildArgs "-p:IntDir=$BinaryCache\$($Platform.Triple)\mimalloc\mimalloc-override-dll\"
 
