diff --git a/stdlib/public/SwiftShims/swift/shims/LibcShims.h b/stdlib/public/SwiftShims/swift/shims/LibcShims.h
index b78e80222f6..c1d3688c1b0 100644
--- a/stdlib/public/SwiftShims/swift/shims/LibcShims.h
+++ b/stdlib/public/SwiftShims/swift/shims/LibcShims.h
@@ -38,6 +38,9 @@ SWIFT_RUNTIME_STDLIB_INTERNAL
 __swift_size_t _swift_stdlib_fwrite_stdout(const void *ptr, __swift_size_t size,
                                            __swift_size_t nitems);
 
+SWIFT_RUNTIME_STDLIB_INTERNAL
+int _swift_stdlib_fflush_stdout();
+
 // General utilities <stdlib.h>
 // Memory management functions
 static inline void _swift_stdlib_free(void *_Nullable ptr) {
diff --git a/stdlib/public/core/Print.swift b/stdlib/public/core/Print.swift
index 015f538bfe0..264c7e1c265 100644
--- a/stdlib/public/core/Print.swift
+++ b/stdlib/public/core/Print.swift
@@ -12,6 +12,10 @@
 
 #if !SWIFT_STDLIB_STATIC_PRINT
 
+#if os(Windows)
+import SwiftShims
+#endif
+
 /// Writes the textual representations of the given items into the standard
 /// output.
 ///
@@ -231,6 +235,9 @@ internal func _print<Target: TextOutputStream>(
     prefix = separator
   }
   output.write(terminator)
+  #if os(Windows)
+  _swift_stdlib_fflush_stdout()
+  #endif
 }
 
 internal func _debugPrint<Target: TextOutputStream>(
@@ -248,6 +255,9 @@ internal func _debugPrint<Target: TextOutputStream>(
     prefix = separator
   }
   output.write(terminator)
+  #if os(Windows)
+  _swift_stdlib_fflush_stdout()
+  #endif
 }
 
 #endif
diff --git a/stdlib/public/stubs/LibcShims.cpp b/stdlib/public/stubs/LibcShims.cpp
index 897f8d0d906..789697a1cd5 100644
--- a/stdlib/public/stubs/LibcShims.cpp
+++ b/stdlib/public/stubs/LibcShims.cpp
@@ -40,3 +40,8 @@ __swift_size_t _swift_stdlib_fwrite_stdout(const void *ptr,
                                            __swift_size_t nitems) {
   return fwrite(ptr, size, nitems, stdout);
 }
+
+SWIFT_RUNTIME_STDLIB_INTERNAL
+int _swift_stdlib_fflush_stdout() {
+    return fflush(stdout);
+}
