diff --git a/include/swift/Basic/LangOptions.h b/include/swift/Basic/LangOptions.h
index 8b400f2d..325a9eba 100644
--- a/include/swift/Basic/LangOptions.h
+++ b/include/swift/Basic/LangOptions.h
@@ -303,6 +303,11 @@ namespace swift {
     /// was not compiled with -enable-testing.
     bool EnableTestableAttrRequiresTestableModule = true;
 
+    /// Experimental
+    /// Disables @objc and dynamic attrs
+    /// Ignored when EnableObjCInterop == true
+    bool DisableObjCAttr = false;
+
     ///
     /// Flags for developers
     ///
diff --git a/include/swift/Option/FrontendOptions.td b/include/swift/Option/FrontendOptions.td
index 8dd0e779..eaf18eb9 100644
--- a/include/swift/Option/FrontendOptions.td
+++ b/include/swift/Option/FrontendOptions.td
@@ -122,6 +122,11 @@ def verify_generic_signatures : Separate<["-"], "verify-generic-signatures">,
 def show_diagnostics_after_fatal : Flag<["-"], "show-diagnostics-after-fatal">,
   HelpText<"Keep emitting subsequent diagnostics after a fatal error">;
   
+def experimental_disable_objc_attr : Flag<["-"], "experimental-disable-objc-attr">,
+  Flags<[FrontendOption]>,
+  HelpText<"Disables parsing @objc and dynamic attrs. "
+           "Ignored if Objective-C interop enabled">;
+
 def enable_cross_import_overlays : Flag<["-"], "enable-cross-import-overlays">,
   HelpText<"Automatically import declared cross-import overlays.">;
 def disable_cross_import_overlays : Flag<["-"], "disable-cross-import-overlays">,
diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index c0ee9217..c13d9ffc 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -233,6 +233,7 @@ void ToolChain::addCommonFrontendArgs(const OutputInfo &OI,
                        options::OPT_disable_actor_data_race_checks);
   inputArgs.AddLastArg(arguments, options::OPT_warn_concurrency);
   inputArgs.AddLastArg(arguments, options::OPT_strict_concurrency);
+  inputArgs.AddLastArg(arguments, options::OPT_experimental_disable_objc_attr);
   inputArgs.AddAllArgs(arguments, options::OPT_enable_experimental_feature);
   inputArgs.AddAllArgs(arguments, options::OPT_enable_upcoming_feature);
   inputArgs.AddLastArg(arguments, options::OPT_warn_implicit_overrides);
diff --git a/lib/Frontend/CompilerInvocation.cpp b/lib/Frontend/CompilerInvocation.cpp
index e07c122e..31876162 100644
--- a/lib/Frontend/CompilerInvocation.cpp
+++ b/lib/Frontend/CompilerInvocation.cpp
@@ -593,6 +593,8 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
 
   Opts.EnableThrowWithoutTry |= Args.hasArg(OPT_enable_throw_without_try);
 
+  Opts.DisableObjCAttr |= Args.hasArg(OPT_experimental_disable_objc_attr);
+
   if (auto A = Args.getLastArg(OPT_enable_objc_attr_requires_foundation_module,
                                OPT_disable_objc_attr_requires_foundation_module)) {
     Opts.EnableObjCAttrRequiresFoundation
diff --git a/lib/Parse/ParseDecl.cpp b/lib/Parse/ParseDecl.cpp
index 9ca39dcc..090a70ba 100644
--- a/lib/Parse/ParseDecl.cpp
+++ b/lib/Parse/ParseDecl.cpp
@@ -5030,6 +5030,29 @@ Parser::parseDecl(ParseDeclOptions Flags,
       return DeclResult;
   }
 
+  if (auto SF = CurDeclContext->getParentSourceFile()) {
+    if (!InInactiveClauseEnvironment) {
+
+      bool ObjCAttrDisabled = Context.LangOpts.DisableObjCAttr && !Context.LangOpts.EnableObjCInterop; // do nothing if ObjCInterop enabled
+
+      bool AttributesChanged = false;
+
+      for (auto Attr : Attributes) {
+        if (isa<ObjCAttr>(Attr)) {
+          if (ObjCAttrDisabled) {
+            Attr->setInvalid();
+            Attributes.removeAttribute(Attr);
+            AttributesChanged = true;
+          }
+        }
+      }
+
+      if (AttributesChanged) {
+        DeclResult.get()->getAttrs() = Attributes;
+      }
+    }
+  }
+
   if (DeclResult.isNonNull()) {
     Decl *D = DeclResult.get();
     if (!HandlerAlreadyCalled)
