diff --git a/utils/build.ps1 b/utils/build.ps1
index f39a6fc551f..bae59ba94ef 100644
--- a/utils/build.ps1
+++ b/utils/build.ps1
@@ -1619,6 +1619,58 @@ function Stage-BuildArtifacts($Arch) {
 
 #-------------------------------------------------------------------
 
+function Build-ToolsPackage($Arch) {
+  $Properties = @{
+    DEVTOOLS_ROOT = "$($Arch.ToolchainInstallRoot)\";
+    TOOLCHAIN_ROOT = "$($Arch.ToolchainInstallRoot)\";
+  }
+
+  Build-WiXProject bld\bld.wixproj -Arch $Arch -Properties $Properties
+  Build-WiXProject cli\cli.wixproj -Arch $Arch -Properties $Properties
+  Build-WiXProject dbg\dbg.wixproj -Arch $Arch -Properties $Properties
+  Build-WiXProject ide\ide.wixproj -Arch $Arch -Properties $Properties
+}
+
+function Build-SDKRuntimePackages($Arch) {
+  foreach ($SDK in $SDKArchs) {
+    $Properties = @{
+      PLATFORM_ROOT = "$($SDK.PlatformInstallRoot)\";
+      SDK_ROOT = "$($SDK.SDKInstallRoot)\";
+      BaseOutputPath = "$($Arch.BinaryCache)\installer\";
+    }
+
+    Build-WiXProject rtl\msi\rtlmsi.wixproj -Arch $SDK -Properties $Properties
+    $Properties["InstallerPlatform"] = "$($Arch.ShortName)"
+    Build-WiXProject sdk\sdk.wixproj -Arch $SDK -Properties $Properties
+  }
+}
+
+function Build-InstallerBundle($Arch) {
+  $Properties = @{
+    BundleFlavor = "offline";
+    BuildProjectReferences = "false";
+  }
+
+  Isolate-EnvVars {
+    Invoke-VsDevShell $Arch
+    $VCRedistInstallerPath = "${env:VCToolsRedistDir}\vc_redist.$($Arch.ShortName).exe"
+    if (Test-Path $VCRedistInstallerPath) {
+      $Properties["VCRedistInstaller"] = $VCRedistInstallerPath
+      $Properties["VSVersion"] = $env:VSCMD_VER
+    }
+  }
+
+  foreach ($SDK in $SDKArchs) {
+    $Properties["INCLUDE_$($SDK.VSName.ToUpperInvariant())_SDK"] = "true"
+    $Properties["PLATFORM_ROOT_$($SDK.VSName.ToUpperInvariant())"] = "$($SDK.PlatformInstallRoot)\"
+    $Properties["SDK_ROOT_$($SDK.VSName.ToUpperInvariant())"] = "$($SDK.SDKInstallRoot)\"
+  }
+
+  Build-WiXProject bundle\installer.wixproj -Bundle -Arch $Arch -Properties $Properties
+}
+
+#-------------------------------------------------------------------
+
 if (-not $SkipBuild) {
   Ensure-WindowsSDK
 }
@@ -1683,7 +1735,9 @@ if (-not $SkipBuild) {
 }
 
 if (-not $SkipPackaging) {
-  Invoke-BuildStep Build-Installer $HostArch
+  Invoke-BuildStep Build-ToolsPackage $HostArch
+  Invoke-BuildStep Build-SDKRuntimePackages $HostArch
+  Invoke-BuildStep Build-InstallerBundle $HostArch
 }
 
 if ($Stage) {
