diff --git a/src/queue.c b/src/queue.c
index 40f059c..7bae11e 100644
--- a/src/queue.c
+++ b/src/queue.c
@@ -6165,7 +6165,8 @@ _dispatch_root_queue_init_pthread_pool(dispatch_queue_global_t dq,
 	dispatch_pthread_root_queue_context_t pqc = dq->do_ctxt;
 	int thread_pool_size = DISPATCH_WORKQ_MAX_PTHREAD_COUNT;
 	if (!(pri & DISPATCH_PRIORITY_FLAG_OVERCOMMIT)) {
-		thread_pool_size = (int32_t)dispatch_hw_config(active_cpus);
+		uint32_t active_cpus = dispatch_hw_config(active_cpus);
+		thread_pool_size = max(64, active_cpus * 5);
 	}
 	if (pool_size && pool_size < thread_pool_size) thread_pool_size = pool_size;
 	dq->dgq_thread_pool_size = thread_pool_size;
